### PR TITLE
Add drive-backed rule draft attachments

### PR DESCRIPTION
**Pull request by elie222**
# User description
Adds drive-backed attachments for rule-generated drafts.

Rules can now approve files and folders as attachment sources, browse them in the UI, and use them during draft generation. The backend adds source discovery, document indexing, caching, and provider support so generated drafts can include selected PDFs from connected drives. Includes schema updates, drive source APIs, and focused tests for the draft attachment flow.

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
Enable rules to persist approved Drive sources, expose a picker in the assistant UI, and route the stored sources through rule persistence so admins can select files/folders approved for attachments. Enhance generator pipelines by indexing approved PDFs, selecting relevant attachments with the AI model, caching those picks per rule, and flowing the resolved files into drafts, replies, and sends so the email providers can upload them.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1879?tool=ast&topic=Email+provider+uploads>Email provider uploads</a>
        </td><td>Surface resolved attachments to outgoing emails by letting Gmail/Outlook providers accept <code>attachments</code>, including new Outlook upload/session logic, MailComposer updates, and ensuring draft/reply/send flows pass sanitized attachments, plus browser flow tests cover Outlook upload edge cases.<details><summary>Modified files (6)</summary><ul><li>apps/web/utils/email/google.ts</li>
<li>apps/web/utils/email/microsoft.ts</li>
<li>apps/web/utils/email/types.ts</li>
<li>apps/web/utils/gmail/mail.ts</li>
<li>apps/web/utils/outlook/mail.test.ts</li>
<li>apps/web/utils/outlook/mail.ts</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>Fix-draft-disable-hand...</td><td>March 14, 2026</td></tr>
<tr><td>josh@jshwrnr.com</td><td>Fix-auto-filer-reply-s...</td><td>February 10, 2026</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1879?tool=ast&topic=Drive+document+indexing>Drive document indexing</a>
        </td><td>Build the backend attachment workflow by indexing Drive files, keeping <code>AttachmentSource</code>/<code>AttachmentDocument</code> tables, looks up attachments per rule using <code>selectDraftAttachmentsForRule</code>, caching selections in redis, and validating/storing attachments while providing targeted tests and crate Gmail/Outlook attachment plumbing.<details><summary>Modified files (20)</summary><ul><li>apps/web/__tests__/eval/draft-attachments.test.ts</li>
<li>apps/web/prisma/migrations/20260311130000_add_attachment_sources/migration.sql</li>
<li>apps/web/prisma/migrations/20260315000000_add_action_static_attachments/migration.sql</li>
<li>apps/web/prisma/schema.prisma</li>
<li>apps/web/utils/ai/actions.test.ts</li>
<li>apps/web/utils/ai/actions.ts</li>
<li>apps/web/utils/ai/assistant/chat.ts</li>
<li>apps/web/utils/ai/choose-rule/choose-args.ts</li>
<li>apps/web/utils/ai/reply/draft-reply.ts</li>
<li>apps/web/utils/ai/types.ts</li>
<li>apps/web/utils/attachments/draft-attachments.ts</li>
<li>apps/web/utils/drive/providers/google.ts</li>
<li>apps/web/utils/drive/providers/microsoft.ts</li>
<li>apps/web/utils/drive/source-items.test.ts</li>
<li>apps/web/utils/drive/source-items.ts</li>
<li>apps/web/utils/drive/types.ts</li>
<li>apps/web/utils/redis/reply.test.ts</li>
<li>apps/web/utils/redis/reply.ts</li>
<li>apps/web/utils/reply-tracker/generate-draft.test.ts</li>
<li>apps/web/utils/reply-tracker/generate-draft.ts</li></ul></details><details><summary>Latest Contributors(0)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1879?tool=ast&topic=Rule+attachment+UX>Rule attachment UX</a>
        </td><td>Enable Drive source discovery APIs (<code>/api/user/drive/source-items*</code>), Client-side hooks (<code>useDriveSourceItems</code>, <code>useDriveSourceChildren</code>), the <code>ActionAttachmentsField</code> picker, <code>ActionSteps</code>, and <code>RuleForm</code> to show approved sources, persist them via <code>handleRuleAttachmentSourceSave</code>, and keep rule validation/action sanitization aware of static attachments and attachment-source inputs.<details><summary>Modified files (23)</summary><ul><li>apps/web/app/(app)/[emailAccountId]/assistant/ActionAttachmentsField.tsx</li>
<li>apps/web/app/(app)/[emailAccountId]/assistant/ActionSteps.tsx</li>
<li>apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx</li>
<li>apps/web/app/(app)/[emailAccountId]/assistant/settings/DraftReplies.tsx</li>
<li>apps/web/app/api/user/drive/source-items/[folderId]/route.ts</li>
<li>apps/web/app/api/user/drive/source-items/route.ts</li>
<li>apps/web/app/api/user/rules/[id]/route.ts</li>
<li>apps/web/hooks/useDriveSourceChildren.ts</li>
<li>apps/web/hooks/useDriveSourceItems.ts</li>
<li>apps/web/utils/action-item.test.ts</li>
<li>apps/web/utils/action-item.ts</li>
<li>apps/web/utils/actions/attachment-sources.ts</li>
<li>apps/web/utils/actions/attachment-sources.validation.ts</li>
<li>apps/web/utils/actions/rule.ts</li>
<li>apps/web/utils/actions/rule.validation.ts</li>
<li>apps/web/utils/attachments/rule.test.ts</li>
<li>apps/web/utils/attachments/rule.ts</li>
<li>apps/web/utils/attachments/source-schema.ts</li>
<li>apps/web/utils/rule/consts.ts</li>
<li>apps/web/utils/rule/rule.ts</li>
<li>apps/web/utils/scheduled-actions/executor.ts</li>
<li>apps/web/utils/scheduled-actions/scheduler.ts</li>
<li>qa/browser-flows/drive-draft-attachment-gmail.md</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>ui-adjust-rule-action-...</td><td>March 15, 2026</td></tr>
<tr><td>joshwerner001@gmail.com</td><td>Remove-conditional-tit...</td><td>December 10, 2025</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/elie222/inbox-zero/1879?tool=ast>(Baz)</a>.

### CODE REVIEW

File: apps/web/utils/attachments/draft-attachments.ts [hunk 1]
```diff
@@ -0,0 +1,836 @@
...
+    if (!driveConnection) continue;
+
+    const provider = await getDriveProvider({
+      driveConnection,
+      providers,
+      logger,
+    });
+    if (!provider) continue;
+
+    const downloaded = await provider.downloadFile(selectedAttachment.fileId);
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/attachments/draft-attachments.ts]: Added per-file error handling around `provider.downloadFile()` in `resolveDraftAttachments`, so one failed download now logs a warning and the loop continues resolving the remaining attachments.

File: apps/web/utils/ai/actions.ts [hunk 2]
```diff
@@ -158,13 +160,38 @@ const draft: ActionFunction<{
...
+        emailAccountId,
+        messageId: email.id,
+        ruleId: executedRule.ruleId,
+      })
+    : null;
+  const attachments = cachedDraft?.attachments?.length
+    ? await resolveDraftAttachments({
+        emailAccountId,
+        userId,
+        selectedAttachments: cachedDraft.attachments,
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/ai/actions.ts]: Added a fallback path in the draft action: if the rule-scoped Redis entry is missing, it now reloads the email account and re-runs attachment selection from the rule before resolving files. I also left a TODO to persist selected draft attachments outside Redis so this can eventually stop depending on re-selection.

File: apps/web/app/api/user/drive/source-items/[folderId]/route.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import prisma from "@/utils/prisma";
+import { withEmailAccount } from "@/utils/middleware";
+import { createDriveProviderWithRefresh } from "@/utils/drive/provider";
+import { SafeError } from "@/utils/error";
+import type { Logger } from "@/utils/logger";
+
+const querySchema = z.object({ driveConnectionId: z.string() });
+export type GetDriveSourceChildrenQuery = z.infer<typeof querySchema>;
+
+export type GetDriveSourceChildrenResponse = Awaited<
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/user/drive/source-items/[folderId]/route.ts:15]: Moved the child-route query validation into the shared `utils/actions/drive.validation.ts` module and switched the route type to `z.infer<typeof getDriveSourceChildrenQuerySchema>`, so it follows the centralized validation pattern now.

File: apps/web/app/api/user/drive/source-items/[folderId]/route.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    provider.listFiles(folderId, { mimeTypes: ["application/pdf"] }),
+  ]);
+
+  return {
+    items: [
+      ...folders.map((folder) => ({
+        id: folder.id,
+        name: folder.name,
+        path: folder.path || folder.name,
+        driveConnectionId: driveConnection.id,
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/user/drive/source-items/[folderId]/route.ts]: Extracted the folder/file -> `DriveSourceItem` mapping into `utils/drive/source-items.ts` and reused it from both source-item routes. I also added a small unit test for the shared helper.

File: apps/web/app/api/user/drive/source-items/route.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      const [folders, files] = await Promise.all([
+        provider.listFolders(undefined),
+        provider.listFiles(undefined, { mimeTypes: ["application/pdf"] }),
+      ]);
+
+      for (const folder of folders) {
+        items.push({
+          id: folder.id,
+          name: folder.name,
+          path: folder.path || folder.name,
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/user/drive/source-items/route.ts]: Extracted the shared `buildDriveSourceItems(...)` helper under `utils/drive/source-items.ts` and switched both `GET` handlers to use it, with a small unit test covering the shared mapping behavior.

File: apps/web/utils/attachments/draft-attachments.ts [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+  emailAccount,
+  emailContent,
+}: {
+  candidates: CandidateDocument[];
+  emailAccount: EmailAccountWithAI;
+  emailContent: string;
+}): Promise<SelectedAttachment[]> {
+  const modelOptions = getModel(emailAccount.user, "economy");
+  const generateObject = createGenerateObject({
+    emailAccount,
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/attachments/draft-attachments.ts]: Added request-scoped logging to attachment selection, passed the existing logger through, and made the selection path non-fatal by returning an empty result on failure. `createGenerateObject` already applies the project retry/fallback behavior internally, so the change here is the extra scoped logging plus graceful fallback instead of bubbling the error up.

File: apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx [hunk 7]
```diff
@@ -155,6 +177,14 @@ export function RuleForm({
         }
       }
 
+      const hasDraftAction = data.actions.some(
+        (action) => action.type === ActionType.DRAFT_EMAIL,
+      );
+      const sourcesToSave = hasDraftAction ? attachmentSources : [];
+      if (!hasDraftAction) {
+        setAttachmentSources([]);
+      }
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx]: Removed the eager `setAttachmentSources([])` path and changed the save flow so attachment sources are only persisted when a `DRAFT_EMAIL` action is still present. If drafts are temporarily disabled, local state now stays intact and we skip the attachment-source write entirely, so existing persisted selections are preserved and failed submits no longer wipe the form state. I also added a small test around the centralized helper covering the skip/success/partial cases.

File: apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx [hunk 8]
```diff
@@ -193,7 +223,21 @@ export function RuleForm({
           });
           if (mutate) mutate();
         } else {
-          toastSuccess({ description: "Saved!" });
+          const attachmentSourceResult = await saveAttachmentSources({
+            emailAccountId,
+            ruleId: res.data.rule.id,
+            attachmentSources: sourcesToSave,
+          });
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx]: Extracted the duplicated attachment-source persistence/toast flow into `apps/web/utils/attachments/rule.ts` as `handleRuleAttachmentSourceSave(...)`, and both the create and update branches now call that helper with branch-specific messages. I added a focused unit test for the helper so future changes to this flow only need to be made in one place.

File: apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx [hunk 9]
```diff
@@ -155,6 +177,14 @@ export function RuleForm({
         }
       }
 
+      const hasDraftAction = data.actions.some(
+        (action) => action.type === ActionType.DRAFT_EMAIL,
+      );
+      const sourcesToSave = hasDraftAction ? attachmentSources : [];
+      if (!hasDraftAction) {
+        setAttachmentSources([]);
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/(app)/[emailAccountId]/assistant/RuleForm.tsx]: Fixed. The submit path no longer clears attachmentSources before the save completes. Instead, it leaves local state untouched on failures and only persists attachment sources when a draft action is still part of the rule, which avoids dropping the user's selected sources from the form state.

File: apps/web/utils/outlook/mail.ts [hunk 10]
```diff
@@ -74,6 +74,15 @@ export async function sendEmailWithHtml(
     logger,
   );
 
+  if (body.attachments?.length) {
+    await addAttachmentsToDraft({
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts:84]: Relevant. Microsoft Graphs simple `/attachments` endpoint does cap file size here, so I added an explicit 3 MB guard before upload and a focused unit test covering the rejection path. I did not add upload-session support in this change; the code now fails fast with a clear error instead of reaching the Graph failure path.

File: apps/web/utils/gmail/mail.ts [hunk 11]
```diff
@@ -50,6 +50,9 @@ export const sendEmailBody = z.object({
   attachments: z.array(zodAttachment).optional(),
 });
 export type SendEmailBody = z.infer<typeof sendEmailBody>;
+type MailSendEmailBody = Omit<SendEmailBody, "attachments"> & {
+  attachments?: Attachment[];
+};
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/gmail/mail.ts]: Relevant. I moved the shared attachment-body widening into `utils/types/mail.ts` as `WithMailerAttachments<T>` and both Gmail and Outlook mail helpers now derive their local mail body type from that shared helper instead of duplicating the shape.

File: apps/web/app/(app)/[emailAccountId]/assistant/ActionSteps.tsx [hunk 12]
```diff
@@ -537,6 +550,19 @@ function ActionCard({
 
   const isNotifySender = actionType === ActionType.NOTIFY_SENDER;
 
+  const supportsAttachments =
+    actionType === ActionType.DRAFT_EMAIL ||
+    actionType === ActionType.REPLY ||
+    actionType === ActionType.SEND_EMAIL;
+  const supportsAiSelectedSources = actionType === ActionType.DRAFT_EMAIL;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/(app)/[emailAccountId]/assistant/ActionSteps.tsx:557]: Relevant. The picker still was not reachable in the AI-draft branch because that branch returned early with muted copy only. I moved the attachment field into a shared `attachmentsField` block and now render it inside the AI-draft card branch as well, so AI-selected sources remain available when the draft body is AI-generated.

File: apps/web/utils/attachments/rule.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  const result = await upsertRuleAttachmentSourcesAction(emailAccountId, {
+    ruleId,
+    sources: attachmentSources,
+  });
+
+  if (result?.serverError || result?.validationErrors) {
+    toastError({ description: partialErrorMessage });
+    return "partial" as const;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/attachments/rule.ts:35]: Relevant. I updated `handleRuleAttachmentSourceSave(...)` to normalize action errors through `getActionErrorMessage(...)` instead of always showing the generic fallback, and extended the unit test to verify the detailed server message is surfaced in the toast.

File: apps/web/utils/outlook/mail.ts [hunk 14]
```diff
@@ -462,3 +502,59 @@ function buildGraphRecipients(
...
+  attachments,
+  logger,
+}: {
+  client: OutlookClient;
+  draftId: string;
+  attachments: Attachment[];
+  logger: Logger;
+}) {
+  if (!draftId) return;
+
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts:520]: Relevant. `getAttachmentContentBuffer` now treats string attachment content as base64 when it decodes canonically, with a UTF-8 fallback for plain text content. That keeps binary attachments intact for Outlook uploads, and I added a focused test that exercises a base64-backed PDF attachment.

File: apps/web/utils/outlook/mail.ts [hunk 15]
```diff
@@ -462,3 +502,59 @@ function buildGraphRecipients(
...
+  for (const attachment of attachments) {
+    const content = getAttachmentContentBuffer(attachment.content);
+    if (!content) continue;
+    assertGraphAttachmentSizeSupported({ attachment, content });
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts]: Relevant. I replaced the blanket 3 MB rejection with split handling in `addAttachmentsToDraft`: files up to 3 MB still use `/attachments`, files between 3 MB and 150 MB now go through `createUploadSession`, and only files above the Graph max are rejected. I also added a focused test covering the upload-session path and re-ran the web CI build.

File: apps/web/utils/outlook/mail.ts [hunk 16]
```diff
@@ -462,3 +504,154 @@ function buildGraphRecipients(
...
+      client
+        .getClient()
+        .api(`/me/messages/${draftId}/attachments/createUploadSession`)
+        .post({
+          AttachmentItem: {
+            attachmentType: "file",
+            name: attachment.filename || "attachment.pdf",
+            contentType: attachment.contentType || "application/octet-stream",
+            size: content.length,
+          },
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts:626]: Relevant. Each upload-session chunk now goes through `withOutlookRetry` instead of a plain one-shot `fetch`, so transient 429/5xx and network failures retry before the upload aborts. I preserved the same per-chunk range semantics and final error message, and added a focused test that forces the first chunk to fail once before succeeding on retry.

File: apps/web/utils/outlook/mail.ts [hunk 17]
```diff
@@ -462,3 +504,154 @@ function buildGraphRecipients(
...
+    start += GRAPH_UPLOAD_CHUNK_SIZE_BYTES
+  ) {
+    const end = Math.min(start + GRAPH_UPLOAD_CHUNK_SIZE_BYTES, content.length);
+    const chunk = content.subarray(start, end);
+    const response = await fetch(uploadUrl, {
+      method: "PUT",
+      headers: {
+        "Content-Length": String(chunk.length),
+        "Content-Range": `bytes ${start}-${end - 1}/${content.length}`,
+      },
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts]: Relevant. The upload-session chunk request now sends `Content-Type: application/octet-stream` alongside the existing length and range headers. The focused Outlook upload-session test also asserts that header is present on the chunk request.

File: apps/web/utils/outlook/mail.ts [hunk 18]
```diff
@@ -462,3 +504,154 @@ function buildGraphRecipients(
...
+  }
+
+  for (
+    let start = 0;
+    start < content.length;
+    start += GRAPH_UPLOAD_CHUNK_SIZE_BYTES
+  ) {
+    const end = Math.min(start + GRAPH_UPLOAD_CHUNK_SIZE_BYTES, content.length);
+    const chunk = content.subarray(start, end);
+    const response = await fetch(uploadUrl, {
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts]: Relevant. This was the same underlying issue as the other retry thread. The chunk upload path now wraps each PUT in `withOutlookRetry`, so transient failures retry before moving to the next chunk, and the focused test covers that by forcing the first chunk upload to return a temporary 503 once.

File: apps/web/utils/outlook/mail.ts [hunk 19]
```diff
@@ -462,3 +504,187 @@ function buildGraphRecipients(
...
+  }
+
+  for (
+    let start = 0;
+    start < content.length;
+    start += GRAPH_UPLOAD_CHUNK_SIZE_BYTES
+  ) {
+    const end = Math.min(start + GRAPH_UPLOAD_CHUNK_SIZE_BYTES, content.length);
+    const chunk = content.subarray(start, end);
+    await withOutlookRetry(
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts]: Relevant. I updated the upload-session path so it no longer treats a retried 416 as a hard failure. Chunk uploads now advance according to `nextExpectedRanges`: successful 202 responses use the server-reported next offset, and a 416 triggers an upload-session status fetch so we can resume from the servers current range instead of replaying the same chunk forever. I also added a focused test that simulates a transient fetch failure followed by a 416 and verifies the upload still completes.

File: apps/web/utils/outlook/mail.test.ts [hunk 20]
```diff
@@ -61,6 +69,226 @@ describe("sendEmailWithHtml", () => {
...
+            content: Buffer.alloc(totalSize),
+            contentType: "application/pdf",
+          },
+        ],
+      },
+      createScopedLogger("outlook-mail-test"),
+    );
+
+    expect(
+      fetchMock.mock.calls.some(([, init]) => init?.method === "GET"),
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.test.ts]: Relevant. I strengthened the 416-resume test so it now asserts the next PUT after the status fetch resumes at the expected byte range rather than only checking that a status request happened. I also added a companion test for the 404/405 fallback path so both resume branches are covered.

File: apps/web/utils/outlook/mail.ts [hunk 21]
```diff
@@ -462,3 +504,239 @@ function buildGraphRecipients(
...
+    const nextStart = getNextExpectedRangeStart(
+      uploadStatus.nextExpectedRanges,
+    );
+    if (typeof nextStart === "number" && nextStart > start) {
+      return nextStart;
+    }
+  }
+
+  const errorText = await response.text();
+  const error = new Error(
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts]: Relevant. `getUploadSessionStatus` now returns `null` for 404/405 instead of aborting the send, and the 416 handler uses that as a fallback signal to continue from local chunk progress when Outlook does not expose server-side ranges. I added a focused test that exercises the 404 status-path and verifies the next PUT resumes at the next chunk offset.

File: apps/web/utils/outlook/mail.ts [hunk 22]
```diff
@@ -462,3 +504,239 @@ function buildGraphRecipients(
...
+
+  if (response.status === 200 || response.status === 201) {
+    return totalSize;
+  }
+
+  if (response.status === 202) {
+    const uploadStatus = (await response.json()) as UploadSessionStatus;
+    return (
+      getNextExpectedRangeStart(uploadStatus.nextExpectedRanges) ?? totalSize
+    );
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/outlook/mail.ts]: Relevant. Partial upload responses no longer default to completion: the upload-session path now treats `200` and `202` as in-progress responses that must provide a usable `nextExpectedRanges` offset, and it throws if that offset is missing. That keeps the loop aligned with the server-reported upload position instead of stopping early.
